packages/core/src/impressions.ts

```diff
@@ -44,40 +45,38 @@ export class DefaultMozAdsImpressionObserver implements MozAdsImpressionObserver
   }
 
   private async recordImpression(placementId: string, impressionUrl?: string | null) {
-    if (impressionUrl) {
-      logger.info(`Impression occurred for placement: ${placementId}`, {
-        type: 'impressionObserver.recordImpression.viewed',
-        placementId: placementId,
+    if (!impressionUrl || !URL.canParse(impressionUrl)) {
```

Comment:
Same pattern as for recordClick, this un-nests the `if-else` (where the `invalid_url_error` used to live on the `else` branch) and instead we'll first check for fallback, then url validity, and if we don't error and exit, proceed along the happy path.